### DISCUSSION TITLE
[Support] 自定义组件疑问

### DISCUSSION THREAD

**Discussion by hooliy** [midwayjs/midway#3565] [Q&A] [answered]
### Describe the problem（描述问题）

想进行模块化开发， 比如主应用是系统管理，现有一个 CMS 模块想以组件的方式开发，然后在主应用中引用。使用 mikro-orm，实体应该定义在 CMS 组件中，在主应用中如何使用？

### Midway Versions（Midway 版本）

midway@latest

**Comment by flyingcrp** [comment]
可以看看这个, https://www.midwayjs.org/docs/component_development 
可以通过component包装一个特定的功能给其它地方使用

**Reply by hooliy** [reply]
尝试使用自定义组件，但是引入mikro-orm后在组件中建立的entity，在主应用中引用组件后就找不到实体了。  所以有这个疑问，主应用与子应用怎么既能共享数据库配置又能做到entity的分离，就像在nestjs框架中， 可以通过在模块中导入
 `
imports: [TypeOrmModule.forFeature([User])]
` 
来实现本模块的功能

**Reply by hooliy** [reply]
![image](https://github.com/midwayjs/midway/assets/3368691/f273b227-a1bf-465e-b6cd-2219220f2193)
组件中mikro配置的entities与主应用中配置的mikro配置会相互覆盖，导致 entities 中实体丢失

**Reply by czy88840616** [reply]
数组合并是覆盖，而不是合并，这个在文档配置章节有明确描述。

至于你说的需求，从原则上来说，组件和主应用设计上应该具有隔离性和独立性，特别是数据源不应耦合，如果希望在主应用使用组件的 entity，那么有几种比较合理的实现：

* 1、组件 数据源和主应用分开，比如主应用叫 default，而组件取另外名字，不过这个和你的需求不一样
* 2、组件导出 entity，由主应用去加载，这样做是合理的，如果 entity 较多，可以导出数组，主应用 concat 即可
* 3、如果希望组件能默默做掉，这个要看初始化之后不同的数据源是否有额外添加的接口，如有，可以在组件初始化中添加

**Reply by hooliy** [reply]
组件：
mikro: { dataSource: { default: { entities: [ Author ], } } },

主应用：
mikro: { dataSource: { default: { entities: [ Book ], } } },

不需要在主应用中使用组件的实体，组件和主应用是独立性。 只是配置中的实体不知道怎么合并，目前是覆盖了。所以项目运行后，只能找到 Book 找不到 Author。

**Reply by hooliy** [reply]
组件中的AuthorController 中调用的 Author 实体丢失。

**Reply by flyingcrp** [reply]
总感觉这个依赖怪怪, 既然都决定以组件的形式使用了不是应该组件暴露action 给使用方调用就行吗? 再把entity暴露出去上层再通过entity直接操作db ?

**Reply by czy88840616** [reply]
组件就是导出东西让上层应用使用的，你这种场景，组件就当是 SDK

```ts
import { customEntities } from '你的组件'

mikro: { dataSource: { default: { entities: [ Book, ...customEntities ], } } },
```

**Comment by hooliy** [comment]
> 再把entity暴露出去上层再通过entity直接操作db ?

不是把entity暴露出去，组件是完全隔离的。有自己的 entity 。主应用导入组件后正常是直接能访问组件中 Controller 操作实体业务的 。现在的问题是在组件中进行entity操作，主应用引入组件后，组件中的实体丢失了。

**Comment by hooliy** [comment]
> 组件就是导出东西让上层应用使用的，你这种场景，组件就当是 SDK
> 
> ```ts
> import { customEntities } from '你的组件'
> 
> mikro: { dataSource: { default: { entities: [ Book, ...customEntities ], } } },
> ```

这样的写法就是主应用与组件耦合太多。 本身所有业务都是在组件中完成 ，主应用引入组件后应该自动包含了组件的业务。不需要再配置数据库实体。
比如在 nestjs 框架中，module 就可以通过  TypeOrmModule.forFeature([User]) 来隔离业务 。主应用不用关心 module 中有哪些实体，只要导入module，主应用就自动包含了module中的业务，包括实体及数据库操作。
在midwayjs的组件中定义的实体，导入到主应用后访问组件中的业务就会报实体丢失。

**Reply by czy88840616** [reply]
你说的是对的，主应用不应该关心 module 中有哪些实体，那为啥要在主应用配置组件的 dataSource ？


`导入到主应用后访问组件中的业务就会报实体丢失。` 是指什么？

**Reply by hooliy** [reply]
举个例子：
组件 m1:
```
controller: m1.controller.ts 
entity: user.entity.ts
```
访问 /m1 查询 user实体

主应用 main：
```
controller: home.controller.ts
entity: book.entity.ts
```
访问 /home 查询 book 实体

以上两个业务是完全分离的，无任何关系。 现在进行组合，在 `main` 的 `configuration.ts `导入 `imports: [m1]`， 预期效果应该是 访问 /m1 ，返回 user 。访问 /home 返回 book。 实际效果为： 访问 /home 可以返回 book，但是访问 /m1 提示 找不到 user 实体

**Reply by czy88840616** [reply]
正常吧，同一个数据源的配置会覆盖的呀，这就等于 user 被book 覆盖了呀。。这就是我说的主应用和组件用了同一个数据源的场景，主应用不用关心 module 中有哪些实体，前提是两个不同的数据源。

**Reply by hooliy** [reply]
那像这种情况如何分离呢？ 比如像做一个自由组合的产品。 有的客户可能需要 A组件，那我应用只需要引入 A 组件就行了，有的客户可能需要 B组件只需要引入 B组件就行了。 
还是说 `同一个数据源的场景` 应该把所有业务集成在一个应用中？

**Reply by czy88840616** [reply]
数据源和 entity 放一起，比如都放组件里，主应用只配连接信息，不需要配 entity

**Reply by hooliy** [reply]
`同一个数据源场景` 如果组件中 `数据源的配置会覆盖`  那组件的业务场景就会很有限。 大部分业务要是提取成组件肯定也是需要操作数据源的吧

**Comment by czy88840616** [comment]
数据源可以直接注入的，并不需要从组件导出，导出 entity 即可。如果真有数据源分离的场景，比如主应用就是要额外添加一些 entity 进去，那么组件应该包装成另外的样子。

比如：

如果你的组件 m1

```ts
export default {
  m1: {
    entities: [xxxx]
  }
}
```

在初始化的地方，组件自己添加到数据源中

```ts
export class Main {

   onReady() {
       const config = configService.getConfiguration()
       const customEntities = configService.getConfiguration('m1.entities')
       config.typeorm.datasources['default']['entities'].push(...customEntities);
   }
}
```

这样能保证组件自己拥有数据源，也具有隔离性。

**Reply by hooliy** [reply]
意思是在主应用中不配置，然后在
A 组件配置 
`mikro: {
    dataSource: {
      A: mikroconfig
    }
  },`

B 组件配置 
`mikro: {
    dataSource: {
      B: mikroconfig
    }
  },`

这样 把 A，B  import 到主应用就可以了对吧？

**Reply by czy88840616** [reply]
不同的数据源，这样是可以的，前提是 A，B 里的 entities 是完全独立，没有关联。

**Reply by hooliy** [reply]
好的，多谢啦！ 才了解到这个框架，还在跟着文档写 demo 中。 以前用 nestjs 比较多。我再试试这两种方式。

**Reply by czy88840616** [reply]
虽然很相似，但是两个框架的底层实现，理念，架构还是有很多差异的。

**Reply by hooliy** [reply]
嗯嗯 ，我再多研究研究！国产框架用起来更顺手一些，还可以很方便的打包成单文件。

**Reply by hooliy** [reply]
这种写法好像也是不行

<img width="606" alt="image" src="https://github.com/midwayjs/midway/assets/3368691/cd3f381d-cd8f-4bb5-a9fe-d653b31cb76f">

**Comment by hooliy** [comment]
> 不同的数据源，这样是可以的，前提是 A，B 里的 entities 是完全独立，没有关联。

很奇怪，这种方式启动时能看到实体已经加载，但是访问时还是提示找不到 User 实体

![1705462118786](https://github.com/midwayjs/midway/assets/3368691/3d83ca10-8703-47af-92d1-61138818eb6d)
<img width="887" alt="image" src="https://github.com/midwayjs/midway/assets/3368691/7b3745f8-95e2-4e22-92ce-a60e32acb299">

**Reply by czy88840616** [reply]
你最好给个 demo。。。我没法从图上猜出来

**Reply by hooliy** [reply]
[demo](https://github.com/hooliy/midway_demo.git)

**Reply by czy88840616** [reply]
像是 mikro 的问题，调试进去 em 走的 this.getContext()，永远返回默认的那个 entityManger。

大概知道了，是 requestScope的问题

**Reply by hooliy** [reply]
要怎么处理一下？

**Reply by czy88840616** [reply]
等我把这段删了就行。https://github.com/midwayjs/midway/pull/3574

**Reply by czy88840616** [reply, marked as answer]
@midwayjs/mikro@3.14.5 拉到这个版本，你的 demo 也已经加到测试里了。https://github.com/midwayjs/midway/tree/main/packages/mikro/test/fixtures/multi-enitymanager/src

**Reply by hooliy** [reply]
@midwayjs/mikro@3.14.5 版本报错。  v3.14.3 不报错

<img width="959" alt="image" src="https://github.com/midwayjs/midway/assets/3368691/0cc26071-79f0-4340-9265-56c798e9dc06">

**Comment by hooliy** [comment]
这里报错了。
<img width="957" alt="image" src="https://github.com/midwayjs/midway/assets/3368691/1a233a2a-5d2b-4b4d-9539-2dfc20d54645">

**Reply by czy88840616** [reply]
多实例要指定数据源名字的，你看示例。

**Reply by hooliy** [reply]
嗯嗯，可以了！多谢！
